University-of-Auckland-SOFTENG761/supervision
# feat(frontend, backend): connect rxdb

# Description
Closes [this](https://trello.com/c/X08Mvev2/19-fe-connect-rxdb-to-backend-endpoint) ticket.

lol:
![image](https://user-images.githubusercontent.com/68878136/189774097-b722067f-09e3-4015-892e-f2b1005c14d5.png)

This is a **LARGE** PR with a lot of potential to break things. It also includes a number of design changes. I'm very open to suggestions and requests for improvement.

### Testing:
1. Launch backend with `npx nx serve backend`
2. Launch frontend with `npx nx serve`
3. Navigate to patient details page
4. Add a student with the "plus" button (NOTE: you cannot delete students at this time)

I would recommend checking that changes are persisted to the database by using the GraphQL playground.
You can check concurrency by opening a second browser, browser profile, or a new browser window in a private browsing mode (multiple tabs do NOT show concurrency and aren't explicitly supported in the current config).

Everything should update concurrently ~except for ENUM fields, which will require you to unload the patient and reload it (page refresh, or select a different patient).~ after debounce delay.

You can simulate going "offline" by shutting off the server, making some changes, then turning it back on.

You can simulate deleting a patient by editing the `deletedAt` field of a patient in the GraphQL playground.

### Changes:
* Added functionality to the "Add Patient" button
* Major refactor for patient-related frontend and backend code
* Added RxDB schema and database for patient info
* Created a hook for accessing patient details
* Implemented RxDB pull and push replication, linked with backend
* Added method for mass-setting patients (required for RxDB push replicaiton)
* Resolved issue with `dateOfBirth` field returning null unexpectedly by adding custom scalar type
* Made all patient fields nullable on backend except for ID, to prevent losing information where a student may accidentally forget to fill in fields

### To do before merging
Please let me know if any of these are out of scope or should be in a separate PR.
- [x] Ensure `deletedAt` works
- [x] Ensure offline functionality works
- [x] Fix NPM vulnerabilities.
- [x] Fix RxDB related failure in CI

### To do in other PRs
- [ ] Ensure `lastUpdatedAt` wins
- [ ] Fix `minUpdatedAt` in the `patientReplicationFeed` query
- [ ] Ensure that only required info is being sent and recieved (optimise network usage)

## Type of change

# How Has This Been Tested?

TODO.
Basic local manual testing.

**Test Configuration**:
* Browser: Chrome (latest)
* Device: Windows Laptop

# Checklist:
